pkg/version/version.go

```diff
@@ -14,68 +14,176 @@
+	}
+	if len(out) == 0 {
+		return nil, nil
+	}
+	return out, nil
+}
+
+// UnmarshalYAML implements the Unmarshaler interface.
+func (v *IstioOperatorVersionCompatibility) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type inStruct struct {
+		OperatorVersion          string `yaml:"operatorVersion"`
+		SupportedIstioVersions   string `yaml:"supportedIstioVersions"`
+		RecommendedIstioVersions string `yaml:"recommendedIstioVersions"`
+	}
+	tmp := inStruct{}
+	if err := unmarshal(&tmp); err != nil {
+		return nil
+	}
+
+	if tmp.OperatorVersion == "" {
+		return fmt.Errorf("operatorVersion must be set")
+	}
+	if tmp.SupportedIstioVersions == "" {
+		return fmt.Errorf("supportedIstioVersions must be set")
+	}
+
+	var err error
+	v.OperatorVersion, err = goversion.NewVersion(tmp.OperatorVersion)
+	if err != nil {
+		return err
+	}
+	v.SupportedIstioVersions, err = goversion.NewConstraint(tmp.SupportedIstioVersions)
+	if err != nil {
+		return err
+	}
+	if tmp.RecommendedIstioVersions != "" {
+		v.RecommendedIstioVersions, err = goversion.NewConstraint(tmp.RecommendedIstioVersions)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// MajorVersion represents a major version.
 type MajorVersion struct {
 	Major uint32
 }
 
+// MinorVersion represents a minor version.
 type MinorVersion struct {
 	MajorVersion
 	Minor uint32
 }
 
+// PatchVersion represents a patch version.
 type PatchVersion struct {
 	MinorVersion
 	Patch uint32
 }
 
+// Version represents a version with an optional suffix.
 type Version struct {
 	PatchVersion
-	Prefix string
 	Suffix string
 }
 
+// NewMajorVersion creates an initialized MajorVersion struct.
 func NewMajorVersion(major uint32) MajorVersion {
 	return MajorVersion{
 		Major: major,
 	}
 }
 
+// NewMinorVersion creates an initialized MinorVersion struct.
 func NewMinorVersion(major, minor uint32) MinorVersion {
 	return MinorVersion{
 		MajorVersion: NewMajorVersion(major),
 		Minor:        minor,
 	}
 }
 
+// NewPatchVersion creates an initialized PatchVersion struct.
 func NewPatchVersion(major, minor, patch uint32) PatchVersion {
 	return PatchVersion{
 		MinorVersion: NewMinorVersion(major, minor),
 		Patch:        patch,
 	}
 }
 
-func NewVersion(prefix string, major, minor, patch uint32, suffix string) Version {
+// NewVersion creates an initialized Version struct.
+func NewVersion(major, minor, patch uint32, suffix string) Version {
 	return Version{
 		PatchVersion: NewPatchVersion(major, minor, patch),
-		Prefix:       prefix,
 		Suffix:       suffix,
 	}
 }
 
+// String implements the Stringer interface.
 func (v MajorVersion) String() string {
 	return fmt.Sprintf("%d", v.Major)
 }
 
+// String implements the Stringer interface.
 func (v MinorVersion) String() string {
 	return fmt.Sprintf("%s.%d", v.MajorVersion, v.Minor)
 }
 
+// String implements the Stringer interface.
 func (v PatchVersion) String() string {
 	return fmt.Sprintf("%s.%d", v.MinorVersion, v.Minor)
```

Comment:
BTW the second argument should be v.Patch?

Reply:
Good catch, thanks.